packages/network-controller/src/NetworkController.ts

```diff
@@ -136,6 +211,110 @@ function isInfuraProviderType(type: string): type is InfuraNetworkType {
   return Object.keys(InfuraNetworkType).includes(type);
 }
 
+/**
+ * Builds an identifier for an Infura network client for lookup purposes.
+ *
+ * @param infuraNetworkOrProviderConfig - The name of an Infura network or a
+ * provider config.
+ * @returns The built identifier.
+ */
+function buildInfuraNetworkClientId(
+  infuraNetworkOrProviderConfig:
+    | InfuraNetworkType
+    | (ProviderConfig & { type: InfuraNetworkType }),
+): BuiltInNetworkClientId {
+  if (typeof infuraNetworkOrProviderConfig === 'string') {
+    return `infura||${infuraNetworkOrProviderConfig}` as const;
+  }
+  return `infura||${infuraNetworkOrProviderConfig.type}` as const;
```

Comment:
How important is it that we use a composite ID?

I'm wondering because I'm hoping that we'll eventually migrate the built-in networks to be more similar to the custom networks (one type of network client, one middleware stack, etc., see https://github.com/MetaMask/core/issues/1279), at which point we presumably won't have a need for this `custom||` prefix on each ID.

And until then, we don't have to worry about ID conflicts. The UUIDs we create for custom networks will never conflict with the short built-in network names.

Perhaps dropping the composite ID now would save us the effort of removing it in a migration later?

Reply:
It'd also make it a bit simpler to access network clients in cases where we don't already have the network client ID. We could do soemthing like `networkClients[providerConfig.type || providerConfig.id]` , no need to construct the ID each time.

Reply:
We don't really need the prefix to be able to tell what type of network client it is, that's true. One benefit to using a prefix that I can think of is that it allows the type of a built-in network client ID to be clearly different from the type of a custom network client ID — and actually, it allows a type to be assigned to the network client ID at all. If we drop the prefix, then the type of a custom network client ID is just `string`. At that point, any kind of network client ID itself, at least from a consumer perspective, will end up being `string`, whether Infura or not.

Maybe that's fine though. Network configuration IDs are just `string`s and we're okay with that. I've made this change here: [9e67774](https://github.com/MetaMask/core/pull/1439/commits/9e67774278ae3433014f286fb8631e3a4c7eddec)

Reply:
Right... yeah this does make type narrowing more challenging. It's still possible for us to narrow from `string` to `InfuraNetworkType` though at least. We'd need to compare with each entry rather than just look at the prefix. This seems OK though.